File: src/Java/adventofcode2018/Day22.java
package adventofcode2018;

import java.awt.Point;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;

/**
 * https://adventofcode.com/2018/day/22
 */
public class Day22 {

    static int depth = 9171;
    static int tx = 7, ty = 721;

    // erosion levels map
    static Map<Point, Integer> map = new HashMap<>();

    static int getLevel(Point p) {
        if (map.containsKey(p)) {
            return map.get(p);
        }
        int level;
        if (p.x == 0) {
            level = p.y * 48271;
        } else if (p.y == 0) {
            level = p.x * 16807;
        } else {
            level = getLevel(new Point(p.x, p.y-1)) * getLevel(new Point(p.x-1, p.y));
        }
        level = (level + depth) %  20183;
        map.put(p, level);
        return level;
    }

    static Map<Point, Integer> distances0 = new HashMap<>();
    static Map<Point, Integer> distances1 = new HashMap<>();
    static Map<Point, Integer> distances2 = new HashMap<>();
    static int getDistance(Point p, int tool) {
        if (tool == 0) return distances0.getOrDefault(p, Integer.MAX_VALUE);
        else if (tool == 1) return distances1.getOrDefault(p, Integer.MAX_VALUE);
        else return distances2.getOrDefault(p, Integer.MAX_VALUE);
    }
    static void putDistance(Point p, int tool, int d) {
        if (tool == 0) distances0.put(p, d);
        else if (tool == 1) distances1.put(p, d);
        else if (tool == 2) distances2.put(p, d);
    }

    public static void main(String[] args) throws Exception {
        // depth = 510;
        // tx = 10;
        // ty = 10;

        int risk = 0;
        for (int i=0; i<=ty; i++) {
            for (int j=0; j<=tx; j++) {
                risk += getLevel(new Point(j, i)) % 3;
            }
        }
        System.err.println(risk);

        Deque<int[]> queue = new ArrayDeque<>();
        queue.add(new int[] {0, 0, 1, 0});  // x, y, tool, time
        int[][] dirs = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        for (int k = 0; k<5_000_000; k++) {
            if (queue.isEmpty()) {
                System.out.println("finished after " + k + " steps");
                break;
            }
            int[] pos = queue.poll();
            Point p = new Point(pos[0], pos[1]);
            int tool = pos[2];
            int time = pos[3];
            if (p.x == tx && p.y == ty) {
                if (tool != 1) time += 7;
                putDistance(p, 1, Math.min(time, getDistance(p, 1)));
                continue;
            }
            if (getDistance(p, tool) < time || getDistance(new Point(tx, ty), 1) <= time)
                continue;
            int tool2 = 3 - (tool + getLevel(p) % 3);
            if (getDistance(p, tool2) > time + 7) {
                putDistance(p, tool2, time + 7);
                queue.add(new int[] {p.x, p.y, tool2, time + 7});
            }
            for (int dir[] : dirs) {
                Point p2 = new Point(p.x + dir[0], p.y + dir[1]);
                if (p2.x < 0 || p2.y < 0 || p2.x > 100 || getLevel(p2) % 3 == tool) {
                    continue;
                }
                if (getDistance(p2, tool) > time + 1) {
                    putDistance(p2, tool, time + 1);
                    queue.add(new int[] {p2.x, p2.y, tool, time + 1});
                }
            }
        }
        System.err.println(getDistance(new Point(tx, ty), 1));
    }
}

File: src/Python/2018/Day22.py

from utils import Input, printResult

# https://adventofcode.com/2018/day/22

depth = 9171
target = (7, 721)

# {point: level} dict
map = {}

def getLevel(p: (int, int)) -> int:
    if p in map: return map[p]
    if p[0] == 0: level = p[1] * 48271
    elif p[1] == 0: level = p[0] * 16807
    else: level = getLevel((p[0], p[1]-1)) * getLevel((p[0]-1, p[1]))
    level = (level + depth) % 20183
    map[p] = level
    return level

distances = [{}, {}, {}]

def getDistance(p, tool) -> int:
    return distances[tool].get(p, 1000000)

def putDistance(p, tool, d):
    distances[tool][p] = d

risk = sum(getLevel((j, i)) % 3 for j in range(target[0] + 1) for i in range(target[1] + 1))
printResult(1, risk)

dirs = [[0, 1], [0, -1], [1, 0], [-1, 0]]
queue = [((0, 0), 1, 0)]  # (x, y), tool, time

import time as tm
starttime = tm.time()

def checkCell(p, tool, time) :
    if time < getDistance(p, tool):
        putDistance(p, tool, time)
        queue.append((p, tool, time))

besttime = 1000000

for k in range(5000000):
    if len(queue) == 0:
        break
    p, tool, time = queue.pop(0)
    # current time not better than already found
    if time > getDistance(p, tool) or time >= besttime:
        continue
    if p == target:
        besttime = min(time if tool == 1 else time + 7, besttime)
        putDistance(p, 1, besttime)
        continue
    # check adjacent cells with a current tool
    for [dx, dy] in dirs:
        p2 = (p[0] + dx, p[1] + dy)
        if p2[0] < 0 or p2[1] < 0 or p2[0] > 100 or getLevel(p2) % 3 == tool:
            continue
        checkCell(p2, tool, time + 1)
    # change a tool
    checkCell(p, 3 - (tool + getLevel(p) % 3), time + 7)

printResult(2, getDistance(target, 1))

print("Finished after %d steps and %.2f sec." % (k, tm.time() - starttime))
